File: dotnet/src/OpenQA.Selenium.Winium/SilverlightOptions.cs
namespace OpenQA.Selenium.Winium
{
    #region using

    using System.Collections.Generic;

    using OpenQA.Selenium;
    using OpenQA.Selenium.Remote;

    #endregion

    /// <summary>
    /// Class to manage options specific to <see cref="WiniumDriver"/> 
    /// wich uses <see href="https://github.com/2gis/winphonedrive">Windows Phone Driver</see>.
    /// </summary>
    public class SilverlightOptions : IWiniumOptions
    {
        #region Constants

        private const string ApplicationPathOption = "app";

        private const string DebugConnectToRunningAppOption = "debugConnectToRunningApp";

        private const string DeviceNameOption = "deviceName";

        private const string InnerPortOption = "innerPort";

        private const string LaunchDelayOption = "launchDelay";

        private const string LaunchTimeoutOption = "launchTimeout";

        #endregion

        #region Fields

        private string applicationPath;

        private bool? debugConnectToRunningApp;

        private string deviceName;

        private int? innerPort;

        private int? launchDelay;

        private int? launchTimeout;

        #endregion

        #region Public Properties

        /// <summary>
        /// Gets or sets the absolute local path to an .xap file to be installed and launched.
        /// This capability is not required if debugConnectToRunningApp is specified.
        /// </summary>
        public string ApplicationPath
        {
            set
            {
                this.applicationPath = value;
            }
        }

        /// <summary>
        /// Gets or sets a value indicating whether debug connect to running app.
        /// If true, then application starting step are skipped.
        /// </summary>
        public bool DebugConnectToRunningApp
        {
            set
            {
                this.debugConnectToRunningApp = value;
            }
        }

        /// <summary>
        /// Gets or sets name of emulator to use for running test. 
        /// Note that this capability is not required, if no device name is specified, 
        /// then a default emulator is used.You can specify only partial name, 
        /// first emulator that starts with specified deviceName will be selected. 
        /// </summary>
        public string DeviceName
        {
            set
            {
                this.deviceName = value;
            }
        }

        /// <summary>
        /// Gets or sets the inner port used to communicate between OuterDriver and InnerDrive (injected into Windows Phone app). 
        /// Required only if non-default port was specified in tested app in AutomationServer.Instance.InitializeAndStart call.
        /// </summary>
        public int InnerPort
        {
            set
            {
                this.innerPort = value;
            }
        }

        /// <summary>
        /// Gets or sets launch delay in milliseconds, to be waited to let visuals to initialize after application launched (after successful ping or timeout). 
        /// Use it if the system running emulator is slow.
        /// </summary>
        public int LaunchDelay
        {
            set
            {
                this.launchDelay = value;
            }
        }

        /// <summary>
        /// Gets or sets maximum timeout in milliseconds, to be waited for application to launch.
        /// </summary>
        public int LaunchTimeout
        {
            set
            {
                this.launchTimeout = value;
            }
        }

        #endregion

        #region Public Methods and Operators

        /// <summary>
        /// Convert options to DesiredCapabilities for Windows Phone Driver 
        /// </summary>
        /// <returns>The DesiredCapabilities for Windows Phone Driver with these options.</returns>
        public ICapabilities ToCapabilities()
        {
            var capabilityDictionary = new Dictionary<string, object>
                                           {
                                               { ApplicationPathOption, this.applicationPath }
                                           };

            if (this.debugConnectToRunningApp.HasValue)
            {
                capabilityDictionary.Add(DebugConnectToRunningAppOption, this.debugConnectToRunningApp);
            }

            if (!string.IsNullOrEmpty(this.deviceName))
            {
                capabilityDictionary.Add(DeviceNameOption, this.deviceName);
            }

            if (this.launchTimeout.HasValue)
            {
                capabilityDictionary.Add(LaunchTimeoutOption, this.launchTimeout);
            }

            if (this.launchDelay.HasValue)
            {
                capabilityDictionary.Add(LaunchDelayOption, this.launchDelay);
            }

            if (!string.IsNullOrEmpty(this.deviceName))
            {
                capabilityDictionary.Add(DeviceNameOption, this.deviceName);
            }

            if (this.innerPort.HasValue)
            {
                capabilityDictionary.Add(InnerPortOption, this.innerPort);
            }

            return new DesiredCapabilities(capabilityDictionary);
        }

        #endregion
    }
}

File: java/src/main/java/org/openqa/selenium/winium/SilverlightOptions.java
package org.openqa.selenium.winium;

import org.openqa.selenium.Capabilities;
import org.openqa.selenium.remote.DesiredCapabilities;

import java.util.HashMap;

/**
 * Class to manage options specific to {@link WiniumDriver}
 * which uses <a href="https://github.com/2gis/winphonedrive">Windows Phone Driver</a>
 */
public class SilverlightOptions implements WiniumOptions {
    private static final String APPLICATION_PATH_OPTION = "app";
    private static final String DEBUG_CONNECT_TO_RUNNING_APP_OPTION = "debugConnectToRunningApp";
    private static final String DEVICE_NAME_OPTION = "deviceName";
    private static final String INNER_PORT_OPTION = "innerPort";
    private static final String LAUNCH_DELAY_OPTION = "launchDelay";
    private static final String LAUNCH_TIMEOUT_OPTION = "launchTimeout";

    private String applicationPath;
    private Boolean debugConnectToRunningApp;
    private String deviceName;
    private Integer innerPort;
    private Integer launchDelay;
    private Integer launchTimeout;

    /**
     * Sets the absolute local path to an .xap file to be installed and launched.
     * This capability is not required if debugConnectToRunningApp is specified.
     * @param applicationPath Absolute local path to an .xap file to be installed and launched.
     */
    public void setApplicationPath(String applicationPath) {
        this.applicationPath = applicationPath;
    }

    /**
     * Sets a value indicating whether debug connect to running app.
     * If true, then application starting step are skipped.
     * @param debugConnectToRunningApp Value indicating whether debug connect to running app.
     */
    public void setDebugConnectToRunningApp(Boolean debugConnectToRunningApp) {
        this.debugConnectToRunningApp = debugConnectToRunningApp;
    }

    /**
     * Sets name of emulator to use for running test.
     * Note that this capability is not required, if no device name is specified,
     * then a default emulator is used. You can specify only partial name,
     * first emulator that starts with specified deviceName will be selected.
     * @param deviceName Name of emulator to use for running test.
     */
    public void setDeviceName(String deviceName) {
        this.deviceName = deviceName;
    }

    /**
     * Sets the inner port used to communicate between OuterDriver and InnerDrive (injected into Windows Phone app).
     * Required only if non-default port was specified in tested app in AutomationServer.Instance.InitializeAndStart call.
     * @param innerPort The inner port used to communicate between OuterDriver and InnerDrive
     */
    public void setInnerPort(Integer innerPort) {
        this.innerPort = innerPort;
    }

    /**
     * Sets launch delay in milliseconds, to be waited to let visuals to initialize
     * after application launched (after successful ping or timeout).
     * Use it if the system running emulator is slow.
     * @param launchDelay Launch delay in milliseconds
     */
    public void setLaunchDelay(Integer launchDelay) {
        this.launchDelay = launchDelay;
    }

    /**
     * Sets maximum timeout in milliseconds, to be waited for application to launch
     * @param launchTimeout Maximum timeout in milliseconds, to be waited for application to launch
     */
    public void setLaunchTimeout(Integer launchTimeout) {
        this.launchTimeout = launchTimeout;
    }

    @Override
    public Capabilities toCapabilities() {
        HashMap<String, Object> capabilityDictionary = new HashMap<String, Object>();
        capabilityDictionary.put(APPLICATION_PATH_OPTION, applicationPath);

        if (debugConnectToRunningApp != null) {
            capabilityDictionary.put(DEBUG_CONNECT_TO_RUNNING_APP_OPTION, debugConnectToRunningApp);
        }

        if ((deviceName != null) && (deviceName.length() > 0)) {
            capabilityDictionary.put(DEVICE_NAME_OPTION, deviceName);
        }

        if (launchTimeout != null) {
            capabilityDictionary.put(LAUNCH_TIMEOUT_OPTION, launchTimeout);
        }

        if (launchDelay != null) {
            capabilityDictionary.put(LAUNCH_DELAY_OPTION, launchDelay);
        }

        if (innerPort != null) {
            capabilityDictionary.put(INNER_PORT_OPTION, innerPort);
        }

        return new DesiredCapabilities(capabilityDictionary);
    }
}
